doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1827 @@
+assumptions system in great detail. It is recommended to read through that
+guide first to understand what the different assumptions mean and how the
+assumptions system works.
+
+The simplest case is a function that always has a given assumption regardless
+of its input. In this case, you can define <code class="docutils
+literal notranslate"><span class="pre">is_*assumption*</span></code> directly
+on the class.
+
+For example, our [example `divides`
+function](custom-functions-divides-definition) is always an integer, because
+its value is always either 0 or 1:
+
+
+```{sidebar} Note
+
+From here on out in this guide, in the interest of space, we will omit the
+previous method definitions in the examples unless they are needed for the
+given example to work. There are [complete
+examples](custom-functions-complete-examples) at the end of this guide with
+all the methods.
+
+```
+
+```py
+>>> class divides(Function):
+...     is_integer = True
+...     is_negative = False
+```
+
+```py
+>>> divides(m, n).is_integer
+True
+>>> divides(m, n).is_nonnegative
+True
+```
+
+In general, however, the assumptions of a function depend on the assumptions
+of its inputs. In this case, you should define an <code class="docutils literal notranslate"><span
+class="pre">\_eval\_*assumption*</span></code> method.
+
+For our [$\operatorname{versin}(x)$
+example](custom-functions-versine-definition), the function is always in $[0,
+2]$ when $x$ is real, and it is 0 exactly when $x$ is an even multiple of
+$\pi$. So `versin(x)` should be *nonnegative* whenever `x` is *real* and
+*positive* whenever `x` is *real* and not an *even* multiple of π. Remember
+that by default, a function's domain is all of $\mathbb{C}$, and indeed
+`versin(x)` makes perfect sense with non-real `x`.
+
+In the assumptions handler methods, as in all methods, we can access the
+arguments of the function using `self.args`.
+
+To see if `x` is an even multiple of `pi`, we can use {meth}`~.as_independent`
+to match `x` structurally as `coeff*pi`. Pulling apart subexpressions
+structurally like this in assumptions handlers is preferable to using
+something like `(x/pi).is_even`, because that will create a new expression
+`x/pi`. The creation of a new expression is much slower. Furthermore, whenever
+an expression is created, the constructors that are called when creating the
+expression will often themselves cause assumptions to be queried. If you are
+not careful, this can lead to infinite recursion. So a good general rule for
+assumptions handlers is, **never create a new expression in an assumptions
+handler**. Always pull apart the args of the function using structural methods
+like `as_independent`.
+
+```py
+>>> from sympy.core.logic import fuzzy_and, fuzzy_not
+>>> class versin(Function):
+...     def _eval_is_nonnegative(self):
+...         # versin(x) is nonnegative iff x is real
+...         x = self.args[0]
+...         return x.is_real
```

Comment:
We need to be more careful here:
```python
In [12]: (1 - cos(pi+I*pi))
Out[12]: 1 + cosh(π)
```
If `x` is not real then that does not imply that `versin(x)` is not nonnegative.

Reply:
Ah good catch. Of course, the real range is just [0, 2], so there are complex values that map it to (2, oo).

Really I would like to just write `(1 - cos(x)).is_real` here. If you look in the guide, a lot of the methods do just delegate to `1 - cos(x)`. Partly that is just me being lazy/keeping the guide simple. But I also do think there are legitimate reasons why you'd want to do that in general. If something can be reduced to something already implemented somewhere else, you'd like to be able to use it. It's generally going to be more robust to reuse something already in SymPy as a user since it's going to be way better tested, and even for library code, it reduces duplication which also minimizes bugs. So it's really unfortunate to have a "no recursive expressions" rule, which I agree is better for performance and avoiding infinite recursion, but makes it harder to avoid rewriting hard mathematical logic.

But also, one of the use-cases for custom functions that this guide talks about is functions that are shorthands for another function (the see FMA example in the guide). In that case, you have a function that represents another expression and evaluates with doit. 

Maybe we just need to have a specific way to create something like this, since the basic pattern is

```py
def Unevaluated(Function):
    @classmethod
    def eval(cls, *args):
        # Evaluate to <expression> on explicit numerical inputs

    def doit(self, deep=True):
        return <expression>

    def _eval_whatever(self):
        return <expression>._eval_whatever()
```

where the last part is basically the same for every possible method that you would want to define. This whole thing can be wrapped up into a single `Function` subclass which you could subclass to specify the expression you want to represent, and only override methods if you don't want the default behavior (like defining a custom code printer).

Reply:
Made https://github.com/sympy/sympy/issues/23701 about the wrapper function superclass idea